src/System.Management.Automation/engine/parser/ast.cs

```diff
@@ -8483,6 +8485,21 @@ public TypeName(IScriptExtent extent, string name, string assembly)
             AssemblyName = assembly;
         }
 
+        /// <summary>
+        /// Construct a typename that represents a generic type definition.
+        /// </summary>
+        /// <param name="extent">The extent of the typename.</param>
+        /// <param name="name">The name of the type.</param>
+        /// <param name="genericArgumentCount">The number of generic arguments.</param>
+        internal TypeName(IScriptExtent extent, string name, int genericArgumentCount)
+            : this(extent, name)
+        {
+            if (genericArgumentCount > 0 && !_name.Contains('`'))
+            {
+                _genericArgumentCount = genericArgumentCount;
+            }
```

Comment:
Usually, we expect a constructor to throw an exception if the argument is unexpected. But here we perform a correction of the argument. It's amazing. I'm afraid this may provoke accidental misuse of this constructor in the future. At the very least, it would be good to add a comment with examples of the scenarios we address and explaining why this correction is needed in this place.  
Alternatively, we could perform a check at the call location and pass the correct arguments.
It could be even better to use a template like `private static TypeName TypeName.CreateAlternativeGenericTypeName()` with a comprehensive description.

Reply:
This "try alternate name" thing happens in multiple places, which is common when dealing with generic type resolution. Comments were added in the `GetReflectionType()` to explain why we only fall back to alternate name even if `genericArgumentCount` is set.

As for the scenario for using this constructor, the XML comment calls it out and it's also easy to understand by looking at where it's used. So, I don't think more comments are needed.

I can add a check to throw when `genericArgumentCount` is a negative value.